src/node.zig

```diff
@@ -331,55 +465,147 @@ fn request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_valu
         translate.throw(env, "Unknown operation.") catch return null;
     }
 
+    const message = context.client.get_message() orelse {
+        translate.throw(
+            env,
+            "Message pool has been exhausted.",
+        ) catch return null;
+    };
+    defer context.client.unref(message);
+
     const operation = @intToEnum(Operation, @intCast(u8, operation_int));
-    const events = decode_slice_from_array(env, argv[2], operation) catch |err| switch (err) {
+    const body_length = decode_events(
+        env,
+        argv[2],
+        operation,
+        message.buffer[@sizeOf(Header)..],
+    ) catch |err| switch (err) {
         error.ExceptionThrown => {
             return null;
         },
     };
-    defer allocator.free(events);
 
-    context.client.request(@bitCast(u128, user_data), on_result, operation, events);
+    context.client.request(@bitCast(u128, user_data), on_result, operation, message, body_length);
 
     return null;
 }
 
-fn on_result(user_data: u128, operation: Operation, results: []const u8) void {
+/// The batch has already been encoded into a byte slice. This means that we only have to do one 
+/// copy directly into an available message. No validation of the encoded data is performed except
+/// that it will fit into the message buffer.
+fn raw_request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_value {
+    var argc: usize = 20;
+    var argv: [20]c.napi_value = undefined;
```

Comment:
I think there are a few places within `node.zig` where we allocate space on the stack like this for `20` argv values. However, we can be more efficient and allocate only as many as we decide we want to get back from `napi_get_cb_info()`. We just set `argc` to the number of args we want, and `argv[N]` to the same, and `napi_get_cb_info()` will only copy that many (if there are more it will stop at how many we expect).

Reply:
For example, in this case, we would want to do:

```zig
var argc: usize = 4;
var argv: [4]c.napi_value = undefined;
```

`napi_get_cb_info()` will then set `argc` to the actual number of args passed by the user, e.g. if more than 4 so we can still do our error handling below for `Function request() requires 4 arguments exactly.` as we do.